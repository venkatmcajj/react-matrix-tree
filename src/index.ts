import Tree from './components/App'

export { Tree }
