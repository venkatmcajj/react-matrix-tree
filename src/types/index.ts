enum Tooltiptype {
  HTML,
  TEXT,
}
export interface IColor {
  color: string
}
export type Arrayvalues = {
  imageSource?: string
  title: string
  tooltip?: string
  tooltiptype?: Tooltiptype
  onClick?: (e: React.MouseEvent<HTMLElement, MouseEvent>, item: Arrayvalues) => void
  data?: Array<Arrayvalues>
}
export type Props = {
  data: Array<Arrayvalues>
  isMobile: boolean
  color: string
}
export type TreeProps = {
  childrens: Array<Arrayvalues>
  level: number
  ismobileview: boolean
  color: string
}
