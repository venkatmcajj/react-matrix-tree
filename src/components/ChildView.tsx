import React from 'react'
import { Arrayvalues, IColor } from '../types'
import { GetTree } from './TreeComponent'
import styled from 'styled-components'

const BottomArrow = styled.div<IColor>`
  border-bottom: 5px solid ${(props) => props.color};
`
const BottomArrowDownContainer = styled.div`
  display: flex;
  justify-content: space-between;
  &:last-child {
    margin-right: -19px;
    margin-left: auto;
  }
`
const TopArrowUpContainer = styled('div')({
  display: 'flex',
  justifyContent: 'center',
})
const BottomArrowDown = styled.div<IColor>`
  content: '';
  width: 0;
  height: 0;
  margin-left: -19px;
  border-top: 15px solid ${(props) => props.color};
  border-left: 20px solid transparent;
  border-right: 20px solid transparent;
  margin-top: 20px;
  position: relative;
  &::before {
    content: '';
    width: 4px;
    height: 25px;
    background: ${(props) => props.color};
    margin-right: -2px;
    margin-bottom: 14px;
    position: absolute;
    bottom: 100%;
    right: 50%;
  }
`
const TopArrowUp = styled.div<IColor>`
  content: '';
  width: 0;
  height: 0;
  border-bottom: 15px solid ${(props) => props.color};
  border-left: 20px solid transparent;
  border-right: 20px solid transparent;
  margin-bottom: 20px;
  position: relative;
  &::before {
    content: '';
    width: 4px;
    height: 25px;
    background: ${(props) => props.color};
    margin-right: -3px;
    margin-top: 14px;
    position: absolute;
    top: 100%;
    right: 50%;
  }
`

type ChildProps = {
  ismobileview: boolean
  level: number
  item: Arrayvalues
  color: string
}
function ChildView({ ismobileview, level, item, color }: ChildProps) {
  const { data = [] } = item
  let canShowChild = false
  if (ismobileview) {
    if (level <= 3 && data.length <= 2) {
      canShowChild = true
    }
  } else if (data.length <= 4) {
    canShowChild = data.length > 0 && level <= 4 ? true : false
  } else {
    canShowChild = false
  }
  return (
    <div>
      {canShowChild && (
        <>
          <div
            style={{
              width: data && data.length > 0 ? `${100 - 100 / data.length}%` : 0,
              marginLeft: 'auto',
              marginRight: 'auto',
              marginBottom: 10,
            }}
          >
            <TopArrowUpContainer>
              <TopArrowUp color={color} />
            </TopArrowUpContainer>
            <BottomArrow color={color} />
            <BottomArrowDownContainer>
              {data &&
                data.map((e1: Arrayvalues, index1: number) => {
                  return <BottomArrowDown color={color} key={`${e1.title}${index1}${level}`}></BottomArrowDown>
                })}
            </BottomArrowDownContainer>
          </div>
          <GetTree color={color} childrens={data} level={level + 1} ismobileview={ismobileview} />
        </>
      )}
    </div>
  )
}

export default ChildView
