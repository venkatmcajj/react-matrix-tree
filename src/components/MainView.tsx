import React from 'react'
import { Arrayvalues } from '../types'
type MainViewTypes = {
  data: Arrayvalues
}
function MainView({ data }: MainViewTypes) {
  return (
    <div>
      {data.imageSource && <img src={data.imageSource} style={{ height: 40, width: 40 }} />}
      <h5
        onClick={(event) => {
          if (data.onClick) {
            data.onClick(event, data)
          }
        }}
        title={data.tooltip}
      >
        {data.title}
      </h5>
    </div>
  )
}

export default MainView
