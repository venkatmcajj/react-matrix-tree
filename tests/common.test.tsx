import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import 'jest-styled-components'
import 'jest-canvas-mock'
import '@testing-library/jest-dom'
import { Tree } from '../src'
import MainView from '../src/components/MainView'
import ChildView from '../src/components/ChildView'

describe('Common render', () => {
  it('Tree renders without crashing', () => {
    render(<Tree color='#6a0136' data={[{ title: 'Message' }]} isMobile={false} />)
  })
})
describe('Child View Render', () => {
  it('Child view render check mobileview', () => {
    const { getAllByText } = render(
      <ChildView
        item={{
          title: 'Test',
          onClick: (e) => {
            e.currentTarget.innerHTML = 'Testing After Click'
          },
          data: [
            {
              title: 'Test',
              data: [
                { title: 'Test', data: [{ title: 'Test', data: [{ title: 'Test', data: [{ title: 'Test' }] }] }] },
              ],
            },
          ],
        }}
        ismobileview={true}
        level={1}
        color='red'
      />,
    )
    expect(getAllByText(/Test/i).length).toBe(3)
  })
})
describe('Child view desktop', () => {
  it('Child view render check desktopview', () => {
    const { getAllByText } = render(
      <ChildView
        item={{
          title: 'Test',
          onClick: (e) => {
            e.currentTarget.innerHTML = 'Testing After Click'
          },
          data: [
            {
              title: 'Test',
              data: [
                { title: 'Test', data: [{ title: 'Test', data: [{ title: 'Test', data: [{ title: 'Test' }] }] }] },
              ],
            },
          ],
        }}
        ismobileview={false}
        level={1}
        color='red'
      />,
    )
    expect(getAllByText(/Test/i).length).toBe(4)
  })
})
describe('Main View render', () => {
  it('Main view changes test after click event', () => {
    const { getByText } = render(
      <MainView
        data={{
          title: 'Test',
          onClick: (e) => {
            e.currentTarget.innerHTML = 'Testing After Click'
          },
        }}
      />,
    )
    fireEvent.click(getByText(/Test/i))
    expect(getByText(/Testing/i)).toBeInTheDocument()
  })
})
