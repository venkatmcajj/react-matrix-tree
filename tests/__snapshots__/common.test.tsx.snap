// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main View render Main view renders without crashing 1`] = `
<div
  onClick={[Function]}
>
  <h5>
    Test
  </h5>
</div>
`;
